### PR TITLE
fix edit on inherited prices

### PR DESCRIPTION
**Pull request by patrickweh**
add searchable to clients

add delete to frontend country list

add default static scope to country

add default country to CreateAddress.php

refactor CreateOrder.php

fix vat_rate_id required even if product is given

add multiple clients to product

change default to net on purchase invoice frontend

fix nested exceptions not being converted to frontend notification

add pivot models to morph map

### CODE REVIEW

File: src/Models/Price.php [hunk 1]
```diff
@@ -39,16 +53,70 @@ protected function casts(): array
...
     {
         return Attribute::get(
             fn () => $this->priceList?->is_net
         );
     }
 
+    public function gross(): Attribute
+    {
+        return Attribute::get(
+            fn () => $this->getGross(data_get($this->product, 'vatRate.rate_percentage') ?: 0)
```
Comment (SirSplasch) [inline src/Models/Price.php:81]: If vatRate is eager loaded, vatRate => vat_rate

File: src/Models/Price.php [hunk 2]
```diff
@@ -39,16 +53,70 @@ protected function casts(): array
...
+        );
+    }
+
+    public function net(): Attribute
+    {
+        return Attribute::get(
+            fn () => $this->getNet(data_get($this->product, 'vatRate.rate_percentage') ?: 0)
```
Comment (SirSplasch) [inline src/Models/Price.php:88]: If vatRate is eager loaded, vatRate => vat_rate

File: src/Models/Price.php [hunk 3]
```diff
@@ -39,16 +53,70 @@ protected function casts(): array
...
+    {
+        return Attribute::get(
+            fn () => $this->discountFlat
+        );
+    }
+
+    public function appliedDiscounts(): Attribute
+    {
+        return Attribute::get(
+            fn () => collect($this->appliedDiscounts)->toArray()
```
Comment (SirSplasch) [inline src/Models/Price.php:116]: what does collect()->toArray() do? I suspect nothing.
Also appliedDiscounts is neither a property nor a relation

File: src/Models/Price.php [hunk 4]
```diff
@@ -12,19 +12,33 @@
 use Illuminate\Database\Eloquent\Relations\BelongsTo;
 use Illuminate\Database\Eloquent\Relations\HasMany;
 use TeamNiftyGmbH\DataTable\Casts\Money;
+use TeamNiftyGmbH\DataTable\Casts\Percentage;
 
 class Price extends Model
 {
     use HasFrontendAttributes, HasPackageFactory, HasUserModification, HasUuid, SoftDeletes;
 
     protected $appends = [
```
Comment (SirSplasch) [inline src/Models/Price.php:21]: appended on every model query

File: src/Models/Price.php [hunk 5]
```diff
@@ -12,19 +12,33 @@
...
+        'discount_flat',
+        'discount_percentage',
+        'is_inherited',
     ];
 
     protected $guarded = [
         'id',
     ];
 
+    protected $with = [
```
Comment (SirSplasch) [inline src/Models/Price.php:36]: eager loaded on every model query

File: src/Models/Price.php [hunk 6]
```diff
@@ -39,16 +53,70 @@ protected function casts(): array
...
+            'discount_flat' => Money::class,
+            'discount_percentage' => Percentage::class,
+            'is_inherited' => 'boolean',
         ];
     }
 
+    public function basePrice(): Attribute
+    {
+        return Attribute::get(
+            fn () => $this->basePrice
```
Comment (SirSplasch) [inline src/Models/Price.php:67]: this is not how it works. basePrice is neither a property nor a relation

File: src/Models/Price.php [hunk 7]
```diff
@@ -39,16 +53,70 @@ protected function casts(): array
...
+    {
+        return Attribute::get(
+            fn () => $this->getNet(data_get($this->product, 'vatRate.rate_percentage') ?: 0)
+        );
+    }
+
+    public function isInherited(): Attribute
+    {
+        return Attribute::get(
+            fn () => $this->isInherited
```
Comment (SirSplasch) [inline src/Models/Price.php:95]: this is not how it works. isInherited is neither a property nor a relation

File: src/Models/Price.php [hunk 8]
```diff
@@ -39,16 +53,70 @@ protected function casts(): array
...
+        );
+    }
+
+    public function discountPercentage(): Attribute
+    {
+        return Attribute::get(
+            fn () => $this->discountPercentage
```
Comment (SirSplasch) [inline src/Models/Price.php:102]: this is not how it works. discountPercentage is neither a property nor a relation

File: src/Models/Price.php [hunk 9]
```diff
@@ -39,16 +53,70 @@ protected function casts(): array
...
+        );
+    }
+
+    public function discountFlat(): Attribute
+    {
+        return Attribute::get(
+            fn () => $this->discountFlat
```
Comment (SirSplasch) [inline src/Models/Price.php:109]: this is not how it works. discountFlat is neither a property nor a relation

**Review by SirSplasch — CHANGES_REQUESTED** [review]

File: src/Helpers/PriceHelper.php [hunk 10]
```diff
@@ -206,7 +207,12 @@ public function price(): ?Price
             $originalPrice = $price->basePrice->{$function}($this->product->vatRate?->rate_percentage);
 
             $price->discountFlat = bcsub($originalPrice, $price->price);
-            $price->discountPercentage = bcdiv($price->price, $originalPrice);
+            $price->discountPercentage = $originalPrice != 0 ? bcdiv($price->price, $originalPrice) : 0;
```
[the remark this answers is not in the document]
Comment (patrickweh) [reply src/Helpers/PriceHelper.php:210]: if $originalPrice is 0 bcdiv throws an divide by zero exception

File: src/Models/Address.php [hunk 11]
```diff
@@ -116,6 +116,43 @@ protected static function booted(): void
                     ->update($addressesUpdates);
             }
         });
+
+        static::deleted(function (Address $address) {
+            $contactUpdates = [];
+            $addressesUpdates = [];
+            $mainAddress = app(Address::class)
```
[the remark this answers is not in the document]
Comment (patrickweh) [reply src/Models/Address.php:123]: main address cant be deleted

File: src/Rulesets/OrderPosition/CreateOrderPositionRuleset.php [hunk 12]
```diff
@@ -84,8 +84,12 @@ public function rules(): array
             'vat_rate_id' => [
                 'exclude_if:is_free_text,true',
                 'exclude_if:is_bundle_position,true',
-                'required_if:is_free_text,false',
-                'required_if:is_bundle_position,false',
+                Rule::when(
```
[the remark this answers is not in the document]
Comment (patrickweh) [reply src/Rulesets/OrderPosition/CreateOrderPositionRuleset.php:87]: tried it in frontend, seems to work

**Review by SirSplasch — APPROVED** [review]

**Pull request merged**